### ISSUE TITLE
Apps crash on Android when activating Now on Tap

### ISSUE THREAD

**Issue by simolus3** [flutter/flutter#11292]
Flutter apps crash on Android when activating the Now on Tap screen search feature of the Google app:
![](https://i.imgur.com/BfkBTgd.gif)

## Steps to Reproduce

This appears to be affecting all Android apps built with fluttter (both in debug and release mode).
```bash
flutter create search_overlay_bug
cd search_overlay_bug
flutter run -v
```
On devices with Google screen search enabled (I don't know if this can be triggered with the assistant as well), long-pressing the home button will open the overlay and then crash the app, as shown above.

## Logs
Outputs of `flutter run -v`: [crash.txt](https://github.com/flutter/flutter/files/1159007/crash.txt)
Outputs of `flutter analyze`:
```
Analyzing /home/simon/search_overlay_bug...
No issues found!
Ran in 6.9s
```
## Flutter Doctor
```
[✓] Flutter (on Linux, locale de_DE.UTF-8, channel alpha)
    • Flutter at /home/simon/bin/flutter
    • Framework revision d36e2f6191 (vor 4 Wochen), 2017-06-21 15:09:10 -0700
    • Engine revision b0dee695ec
    • Tools Dart version 1.24.0-dev.6.7

[✓] Android toolchain - develop for Android devices (Android SDK 25.0.3)
    • Android SDK at /home/simon/Android/Sdk
    • Platform android-25, build-tools 25.0.3
    • Java binary at: /opt/android-studio/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_112-release-b06)

[✓] Android Studio (version 2.3)
    • Android Studio at /opt/android-studio
    • Gradle version 3.2
    • Java version OpenJDK Runtime Environment (build 1.8.0_112-release-b06)

[✓] IntelliJ IDEA Community Edition (version 2017.1)
    • Flutter plugin version 15.1
    • Dart plugin version 171.4694.29

[✓] Connected devices
    • MotoG3 • ZY22346LJL • android-arm • Android 6.0.1 (API 23)
```

**Comment by eseidelGoogle** [comment]
We don't seem to provide very much info in the crash dump:
```
[    +3738 ms] I/AssistStructure(  496): Flattened final assist data: 804 bytes, containing 1 windows, 8 views
[     +420 ms] F/flutter (  496): [FATAL:flutter/shell/platform/android/platform_view_android_jni.cc(49)] Check failed: env->ExceptionCheck() == JNI_FALSE. 
[            ] F/libc    (  496): Fatal signal 6 (SIGABRT), code -6 in tid 496 (earchoverlaybug)
[     +696 ms] Service protocol connection closed.
```

FYI @goderbauer @jason-simmons

**Comment by Hixie** [comment]
This may be fixed on master; can you try switching to the master channel and seeing if it works?

**Comment by simolus3** [comment]
Thank you, it is fixed on master. However, FlutterView logs the following exception when opening the overlay:
```
[+5018 ms] I/AssistStructure(26288): Flattened final assist data: 804 bytes, containing 1 windows, 8 views


[ +340 ms] E/FlutterView(26288): Uncaught exception while updating semantics
[        ] E/FlutterView(26288): java.lang.IllegalStateException: Accessibility off. Did you forget to check that?
[   +2 ms] E/FlutterView(26288): 	at android.view.accessibility.AccessibilityManager.sendAccessibilityEvent(AccessibilityManager.java:309)
[        ] E/FlutterView(26288): 	at android.view.ViewRootImpl.requestSendAccessibilityEvent(ViewRootImpl.java:6426)
[        ] E/FlutterView(26288): 	at android.view.ViewGroup.requestSendAccessibilityEvent(ViewGroup.java:880)
[        ] E/FlutterView(26288): 	at android.view.ViewGroup.requestSendAccessibilityEvent(ViewGroup.java:880)
[        ] E/FlutterView(26288): 	at android.view.ViewGroup.requestSendAccessibilityEvent(ViewGroup.java:880)
[        ] E/FlutterView(26288): 	at io.flutter.view.AccessibilityBridge.sendAccessibilityEvent(AccessibilityBridge.java:313)
[        ] E/FlutterView(26288): 	at io.flutter.view.AccessibilityBridge.updateSemantics(AccessibilityBridge.java:299)
[        ] E/FlutterView(26288): 	at io.flutter.view.FlutterView.updateSemantics(FlutterView.java:689)
[        ] E/FlutterView(26288): 	at android.os.MessageQueue.nativePollOnce(Native Method)
[        ] E/FlutterView(26288): 	at android.os.MessageQueue.next(MessageQueue.java:323)
[        ] E/FlutterView(26288): 	at android.os.Looper.loop(Looper.java:135)
[        ] E/FlutterView(26288): 	at android.app.ActivityThread.main(ActivityThread.java:5443)
[        ] E/FlutterView(26288): 	at java.lang.reflect.Method.invoke(Native Method)
[        ] E/FlutterView(26288): 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:728)
[        ] E/FlutterView(26288): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:618)

```

**Comment by Hixie** [comment]
cc @goderbauer

**Comment by Hixie** [comment]
This is on our medium term radar. We should check to see if it's fixed, first. :-)

**Comment by goderbauer** [comment]
I can confirm that the exception is still happening.

**Comment by goderbauer** [comment]
Here's what's happening behind the scene: When Assistant is activated with long press on Home, the assistant wants to know what's on screen (to offer context-aware help). For that, it calls FlutterView's `getAccessibilityNodeProvider` method and the expectation is that it gives Android a one-off description of what's on screen. However, in Flutter we currently interpret this one-off call as "OMG, somebody turned on accessibility. Let's fire up the accessibility pipeline and tell Android about every future semantic change on screen!!1" (see [1]). The first time we tell Android about one of those updates in semantics, Android is like "What are you doing, dude? Accessibility is off, no need to spam me with accessibility events" and throws an exception.

So, to really fix this and to give Assistant a glimpse into what is currently shown in the Flutter app we would have to add support for providing one-off snapshots of the current semantics tree to Android. Given that the API for requesting these one-offs is synchronous and we would have to get the current semantic snapshot asynchronously from Dart land, this will be difficult to do (short of always keeping an updated semantics tree around in the engine in case somebody ask for it, which would be bad for performance). 😞 

The short term solution might be to just respond to Android's one-off requests with "Sorry, there is nothing on screen currently". That would probably be fine for the assistant. Not sure how it would impact other apps that request one-off descriptions of the current screen content (passwords managers, maybe?).

[1] https://github.com/flutter/engine/blob/6bacf4522e36550b017a5fd43c202aa32d6b9273/shell/platform/android/io/flutter/view/FlutterView.java#L857

**Comment by Hixie** [comment]
We should contact the assistant team and get their take on this. We can't be the only team that isn't set up to give a synchronous answer to this API.

**Comment by eseidelGoogle** [comment]
Chrome likely has a similar problem, since their renderers are in another process. :)  Might be able to find some inspiration there.
